data/transactions/logic/resources_test.go

```diff
@@ -0,0 +1,887 @@
+			logic.NewExpect(1, "invalid Asset reference 202"))
+		// And adding asa2 does not fix this problem, because the other x 202 holding is unavailable
+		appl.ForeignAssets = []basics.AssetIndex{asa2}
+		logic.TestApps(t, []string{"", axferToArgs}, txntest.Group(&appl0, &appl), 9, ledger,
+			logic.NewExpect(1, "axfer AssetReceiver: invalid Holding access "+otherAcct.String()+" x 202"))
+
+		// Now, conduct similar tests, but with the apps performing the
+		// pays/axfers invoked from an outer app. Use various versions to check
+		// cross version sharing.
+
+		// add v8 and v9 versions of the pay app to the ledger for inner calling
+		payToArgV8 := logic.TestProg(t, payToArg, 8)
+		ledger.NewApp(senderAcct, 88, basics.AppParams{ApprovalProgram: payToArgV8.Program})
+		ledger.NewAccount(appAddr(88), 1_000_000)
+		payToArgV9 := logic.TestProg(t, payToArg, 9)
+		ledger.NewApp(senderAcct, 99, basics.AppParams{ApprovalProgram: payToArgV9.Program})
+		ledger.NewAccount(appAddr(99), 1_000_000)
+
+		approvalV8 := logic.TestProg(t, "int 1", 8)
+		ledger.NewApp(senderAcct, 11, basics.AppParams{ApprovalProgram: approvalV8.Program})
+
+		innerCallTemplate := `
+itxn_begin
+int appl;                     itxn_field TypeEnum;
+txn ApplicationArgs 0; btoi;  itxn_field ApplicationID
+txn ApplicationArgs 1;        itxn_field ApplicationArgs
+txn ApplicationArgs 2;        itxn_field ApplicationArgs
+%s
+itxn_submit
+int 1
+`
+		innerCall := fmt.Sprintf(innerCallTemplate, "")
+
+		appl.ForeignApps = []basics.AppIndex{11, 88, 99}
+
+		appl.ApplicationArgs = [][]byte{{99}, otherAcct[:], {asa1}}
+		logic.TestApps(t, []string{"", innerCall}, txntest.Group(&appl0, &appl), 9, ledger)
+		// when the inner program is v8, it can't perform the pay
+		appl.ApplicationArgs = [][]byte{{88}, otherAcct[:], {asa1}}
+		logic.TestApps(t, []string{"", innerCall}, txntest.Group(&appl0, &appl), 9, ledger,
+			logic.NewExpect(1, "invalid Account reference "+otherAcct.String()))
+		// unless the caller passes in the account, but it can't because that
+		// would give the called app access to the passed account's locate state
+		innerCallWithAccount := fmt.Sprintf(innerCallTemplate, "addr "+otherAcct.String()+"; itxn_field Accounts")
+		logic.TestApps(t, []string{"", innerCallWithAccount}, txntest.Group(&appl0, &appl), 9, ledger,
+			logic.NewExpect(1, "appl ApplicationID: invalid Local State access "+otherAcct.String()))
+		// the caller can't fix by passing 88 as a foreign app, because doing so
+		// gives the v8 app access to that local state that the caller does not
+		// have access to.
+		innerCallWithBoth := fmt.Sprintf(innerCallTemplate,
+			"addr "+otherAcct.String()+"; itxn_field Accounts; int 88; itxn_field Applications")
+		logic.TestApps(t, []string{"", innerCallWithBoth}, txntest.Group(&appl0, &appl), 9, ledger,
+			logic.NewExpect(1, "appl ApplicationID: invalid Local State access "+otherAcct.String()))
```

Comment:
I don't understand why `innerCallWithBoth` is interesting here. It seems like this by definition should behave exactly the same as `innerCallWithAccount`. After all, the app being called is 88, so setting `int 88; itxn_field Applications` shouldn't be consequential.

Though I get why it's useful for the final test, where the app being called is 11, so the cross product with 88 is new.

Reply:
Yeah, maybe it's redundant. I guess I was imagining a bug where the appid is allowed in the foreign array and that somehow lets the account pass happen, since it's explicit. As you say, the better test is when the called app is not 88. Do you think I should remove this?

Reply:
Got it. I think the test should stay, but I'd suggest shortening the comment to just "the caller can't fix by passing 88 as a foreign app."

The extended description was what confused me, since there should be no new effects from adding the foreign app 88.

Reply:
addresses